### PR TITLE
Added support for passing http query parameters

### PR DESCRIPTION
**Pull request by Hakky54**
This PR is an implementation for this issue: https://github.com/elastic/elasticsearch/issues/53846

It will give the end user the capability of supplying custom http query parameters to the RestHighLevelClient. This option was already possible with the LowLevelRestClient.

So why do we want this option in the RHLC? This client has useful features such as mapping the response to a SearchResponse. When using the LLRC the end user needs to add custom mappers to generate the SearchResponse object out of the http entity. So for mapping a json back to a SearchResponse with aggregations, suggestions the end user will need the different `NamedXContentRegistry `

The end user may even end up copying the whole content of [RestHighLevelClient#getDefaultNamedXContents()](https://github.com/elastic/elasticsearch/blob/db4ccb6d0792df23bcc37d2328d974f13902f92b/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L1938)

**Below is the original content of the issue mentioned earlier to give some additional context to this PR:**

**Describe the feature**:
_Support for setting url query paramters within the RequestOptions object_

Recently I came to a point where I wanted a smaller search response. The reason why I wanted it was because of network latency, big search responses and not needed fields within the search response. By discarding specific fields the response would be smaller measured in kb.

This option is unfortunately not possible with the RestHighLevelClient and also not possible to add as a property to a SearchRequest. And adding that option to the SearchRequest would mean a change for the specification of the SearchRequest. But I have the feeling that something like that is not a part of the search request.

If we grab an example which I have used at kibana, it would look like this:
```
GET twitter/_search?filter_path=-hits.hits._index
{
  "query": {
    "match_all": {}
  }
}
```

Lets translate this request to a basic search request with the RestHighLevelClient (without filter path):
```
RestClientBuilder clientBuilder = RestClient.builder(new HttpHost("localhost", 9200));
RestHighLevelClient client = new RestHighLevelClient(clientBuilder);

SearchRequest searchRequest = new SearchRequest();
SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()
    .query(new MatchAllQueryBuilder());
searchRequest.source(sourceBuilder)
    .indices("twitter");

SearchResponse response = client.search(searchRequest, RequestOptions.DEFAULT);
```
Let's translate this request to a request to execute with the LowLevelClient including filter_path while preserving the capability of sending a SearchRequest and getting back a SearchResponse
```
RestClientBuilder clientBuilder = RestClient.builder(new HttpHost("localhost", 9200));
RestHighLevelClient client = new RestHighLevelClient(clientBuilder);

SearchRequest searchRequest = new SearchRequest();
SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()
    .query(new MatchAllQueryBuilder());
searchRequest.source(sourceBuilder)
    .indices("twitter");

Request request = new Request(HttpPost.METHOD_NAME, "/" + String.join(",", searchRequest.indices()) + "/_search");
request.setOptions(RequestOptions.DEFAULT);
request.addParameter("filter_path","-hits.hits._index");

HttpEntity httpEntity = new NStringEntity(searchRequest.source().toString(), ContentType.APPLICATION_JSON);
request.setEntity(httpEntity);

Response response = client.getLowLevelClient().performRequest(request)

List<NamedXContentRegistry.Entry> entries = new ArrayList<>();
entries.add(new NamedXContentRegistry.Entry(Aggregation.class, new ParseField(StringTerms.NAME), (parser, content) -> ParsedStringTerms.fromXContent(parser, (String) content)));
entries.add(new NamedXContentRegistry.Entry(Aggregation.class, new ParseField(TopHitsAggregationBuilder.NAME), (parser, content) -> ParsedTopHits.fromXContent(parser, (String) content)));
entries.add(new NamedXContentRegistry.Entry(Suggest.Suggestion.class, new ParseField("term"), (parser, content) -> TermSuggestion.fromXContent(parser, (String) content)));
entries.add(new NamedXContentRegistry.Entry(Suggest.Suggestion.class, new ParseField("phrase"), (parser, content) -> PhraseSuggestion.fromXContent(parser, (String) content)));

NamedXContentRegistry namedXContentRegistry = new NamedXContentRegistry(entries);

String content = EntityUtils.toString(response.getEntity());
XContentParser parser = JsonXContent.jsonXContent.createParser(namedXContentRegistry, null, content);
SearchResponse searchResponse = SearchResponse.fromXContent(parser);
```
This would return me all documents from the twitter index without the index name within the hits.
As you can see there is a lot code needed to just use the filter_path while maintaining the SearchRequest and SearchResponse at a highlevel of our code base.

It would be great if there would be support or a new feature to enable RequestOptions to also support query parameter. Now the RequestOptions just only supports headers. Request options with query parameters could look like this:
```
RequestOptions.DEFAULT.toBuilder()
    .addParameter("filter_path","-hits.hits._index");
```

### CODE REVIEW

**Comment by Hakky54** [comment]
Hi @swallez 

I ended up having the same challenge for a different project. So I created this PR to have it working with the RestHighLevelClient. Almost a year has been passed when I created the issue, so I thought to work on it to help you guys a bit. I know this issue is probably not the one with highest priority on your list.

**Comment by jakelandis** [comment]
@elasticmachine ok to test

@swallez - Can you take a look at this PR ?

**Review by swallez — APPROVED** [review]
LGTM.

Thanks @Hakky54! Very useful addition, and I like the fact that `addParameter` (and not `set`) builds a comma separated string, which is used widely in Elasticsearch.

**Comment by Hakky54** [comment]
Awesome @swallez! Thank you very much for quickly reviewing and approving this PR. Looking forward to the next release! ☺️ it was fun to contribute

**Comment by swallez** [comment]
Thank _you_ @Hakky54! This was of great quality and deserved to be approved right away! Keep them coming 😉

File: client/rest-high-level/src/test/java/org/elasticsearch/client/HighLevelRestClientFilterPathIT.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        doc.setJsonEntity(SAMPLE_DOCUMENT);
+        client().performRequest(doc);
+        client().performRequest(new Request(HttpPost.METHOD_NAME, "/_refresh"));
+
+        RequestOptions requestOptions = RequestOptions.DEFAULT.toBuilder()
+            .addParameter(FILTER_PATH_PARAM, FILTER_PATH_PARAM_VALUE)
+            .build();
+
+        SearchRequest searchRequest = new SearchRequest("company_one");
+        SearchResponse searchResponse = execute(searchRequest, highLevelClient()::search, highLevelClient()::searchAsync, requestOptions);
```
Comment (skumar2proterra) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/HighLevelRestClientFilterPathIT.java:47]: how can we use execute

**Pull request merged**